src/Interpreters/RowRefs.h

```diff
@@ -141,123 +143,157 @@ struct RowRefList : RowRef
  * After calling any of the lookup methods, it is no longer allowed to insert more data as this would invalidate the
  * references that can be returned by the lookup methods
  */
+struct SortedLookupVectorBase
+{
+    SortedLookupVectorBase() = default;
+    virtual ~SortedLookupVectorBase() { }
+
+    static std::optional<TypeIndex> getTypeSize(const IColumn & asof_column, size_t & type_size);
+
+    // This will be synchronized by the rwlock mutex in Join.h
+    virtual void insert(const IColumn &, const Block *, size_t) = 0;
 
-template <typename TEntry, typename TKey>
-class SortedLookupVector
+    // This needs to be synchronized internally
+    virtual std::tuple<decltype(RowRef::block), decltype(RowRef::row_num)> findAsof(const IColumn &, size_t) = 0;
+};
+
+template <typename TKey, ASOF::Inequality inequality>
+class SortedLookupVector : public SortedLookupVectorBase
```

Comment:
I suppose we can hide this into .cpp file, because it is implementation detail.

Reply:
That's a great point :ok_hand: